doc/specs/stdlib_selection.md

```diff
@@ -0,0 +1,340 @@
+The user can optionally specify `left` and `right` indices to constrain the search
+for the kth-smallest value. This can be useful if you have previously called `arg_select`
+to find a smaller or larger rank (that will have led to partial sorting of
+`index`, thus implying some constraints on the location).
+
+#### Licensing
+
+The Fortran Standard Library is distributed under the MIT
+License. However components of the library may be based on code with
+additional licensing restrictions. In particular `select` and `arg_select`
+were derived using some code from quickSelect in the Coretran library, by Leon Foks,
+https://github.com/leonfoks/coretran. Leon has given permission for the code here
+to be released under stdlib's MIT licence.
+
+### Specifications of the `stdlib_selection` procedures
+
+#### `select` - find the kth smallest value in an input array
+
+##### Status
+
+Experimental
+
+##### Description
+
+Returns the k-th smallest value of `array(:)`, and also partially sorts `array(:)`
+such that `all(array(1:k) <= array(k))`  and `all(array(k) <= array((k+1):size(array)))`
+
+##### Syntax
+
+`call [[stdlib_selection(module):select(interface)]]( array, k, kth_smallest [, left, right ] )`
+
+##### Class
+
+Generic subroutine.
+
+##### Arguments
+
+`array` : shall be a rank one array of any of the types:
+`integer(int8)`, `integer(int16)`, `integer(int32)`, `integer(int64)`,
+`real(sp)`, `real(dp)`, `real(qp)`. It is an `intent(inout)` argument. 
+
+`k`: shall be a scalar with any of the types:
+`integer(int8)`, `integer(int16)`, `integer(int32)`, `integer(int64)`. It
+is an `intent(in)` argument. We search for the `k`-th smallest entry of `array(:)`.
+
+`kth_smallest`: shall be a scalar with the same type as `array`. On return it contains
+the k-th smallest entry of `array(:)`.
+
+`left` (optional): shall be a scalar with the same type as `k`. It is an `intent(in)`
+argument. If specified then we assume the k-th smallest value is definitely contained
+in `array(left:size(array))`. If not present it is 1. This is typically useful if multiple calls
+to `select` are made, because the partial sorting of `array` implies constraints on where
+we need to search.
+
+`right` (optional): shall be a scalar with the same type as `k`. It is an `intent(in)`
+argument. If specified then we assume the k-th smallest value is definitely contained
+in `array(1:right)`. If not present it is `size(array)`. This is typically useful if multiple calls
+to `select` are made, because the partial sorting of `array` implies constraints on where
+we need to search.
+
+##### Notes
+
+Selection of a single value should have runtime of O(`size(array)`), so it is
+asymptotically faster than sorting `array` entirely. The test program at the the
+end of this document shows that is the case.
+
+On return the elements of `array` will be partially sorted such that:
+`all( array(1:k-1) <= array(k) )` and `all(array(k) <= array(k+1:size(array)))`.
```

Comment:
Already mentioned in the description. Therefore I think it could be removed.

Reply:
agree, implemented in upcoming commits.